src/main/cpp/lang/xml.cpp

```diff
@@ -0,0 +1,478 @@
+            (str[i + 1] == u8'm' || str[i + 1] == u8'M') && //
+            (str[i + 2] == u8'l' || str[i + 2] == u8'L')) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+} // namespace
+
+[[nodiscard]]
+std::size_t match_whitespace(std::u8string_view str)
+{
+    const std::size_t result = utf8::find_if_not(str, is_xml_whitespace_character);
+
+    return result == std::string::npos ? str.length() : result;
+}
+
+[[nodiscard]]
+Name_Match_Result match_name_permissive(
+    std::u8string_view str,
+    const std::function<bool(std::u8string_view)>& is_stop_sequence
+)
+{
+    if (str.empty()) {
+        return {};
+    }
+
+    Name_Match_Result result;
+
+    if (!is_xml_name_start_character(str[0])) {
+        result.error_indicies.insert(0);
+    }
+
+    result.length++;
+    while (result.length < str.length() && !is_stop_sequence(str.substr(result.length))) {
+        if (!is_xml_name_character(str[result.length])) {
+            result.error_indicies.insert(result.length);
+        }
+
+        result.length++;
+    }
+
+    return result;
+}
+
+[[nodiscard]]
+std::size_t match_text(std::u8string_view str)
+{
+    const std::size_t result
+        = utf8::find_if(str, [](char32_t c) { return c == u8'<' || c == u8'&'; });
+
+    return result == std::string::npos ? str.length() : result;
+}
+
+[[nodiscard]]
+html::Match_Result match_comment(std::u8string_view str)
+{
+    if (!str.starts_with(comment_prefix)) {
+        return {};
+    }
+    str.remove_prefix(comment_prefix.length());
+
+    std::size_t length = comment_prefix.length();
+    while (!str.empty()) {
+
+        if (str.starts_with(comment_suffix)) {
+            return { length + comment_suffix.length(), true };
+        }
+
+        if (str.starts_with(illegal_comment_sequence)) {
+            return { length, false };
+        }
+
+        length++;
+        str.remove_prefix(1);
+    }
+
+    return { length, false };
+}
+
+struct XML_Highlighter : Highlighter_Base {
+
+private:
+    void highlight_name(const Name_Match_Result& result, Highlight_Type type)
+    {
+        std::size_t distance = 0;
+        for (std::size_t i = 0; i < result.length; i++) {
+
+            if (result.error_indicies.contains(i)) {
+
+                if (distance) {
+                    emit_and_advance(distance, type);
+                }
+
+                if (!remainder.empty()) {
+                    emit_and_advance(1, Highlight_Type::error);
+                }
+
+                distance = 0;
+                continue;
+            }
+
+            distance++;
+        }
+
+        if (distance) {
+            emit_and_advance(distance, type);
+        }
+    }
+
+public:
+    XML_Highlighter(
+        Non_Owning_Buffer<Token>& out,
+        std::u8string_view source,
+        const Highlight_Options& options
+    )
+        : Highlighter_Base(out, source, options)
+    {
+    }
+
+    // TODO: add prolog (declaration)
+    bool highlight()
```

Comment:
```suggestion
    bool operator()()
```
The "entry point" for highlighters should be a call operator.